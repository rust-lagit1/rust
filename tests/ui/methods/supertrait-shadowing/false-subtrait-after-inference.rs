#![feature(supertrait_item_shadowing)]

struct W<T>(T);

trait Upstream {
    fn hello(&self) {}
}
impl<T> Upstream for T {}

trait Downstream: Upstream {
    fn hello(&self) {}
}
impl<T> Downstream for W<T> where T: Foo {}

trait Foo {}

fn main() {
    let x = W(Default::default());
    x.hello();
    //~^ ERROR the trait bound `i32: Foo` is not satisfied
    //~| WARN trait method `hello` from `Downstream` shadows identically named method from supertrait
    let _: i32 = x.0;
}
