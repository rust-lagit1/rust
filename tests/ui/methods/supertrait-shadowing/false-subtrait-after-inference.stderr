warning: trait method `hello` from `Downstream` shadows identically named method from supertrait
  --> $DIR/false-subtrait-after-inference.rs:19:7
   |
LL |     x.hello();
   |       ^^^^^
   |
note: method from `Downstream` shadows a supertrait method
  --> $DIR/false-subtrait-after-inference.rs:11:5
   |
LL |     fn hello(&self) {}
   |     ^^^^^^^^^^^^^^^
note: method from `Upstream` is shadowed by a subtrait method
  --> $DIR/false-subtrait-after-inference.rs:6:5
   |
LL |     fn hello(&self) {}
   |     ^^^^^^^^^^^^^^^
   = note: `#[warn(supertrait_item_shadowing)]` on by default

error[E0277]: the trait bound `i32: Foo` is not satisfied
  --> $DIR/false-subtrait-after-inference.rs:19:7
   |
LL |     x.hello();
   |       ^^^^^ the trait `Foo` is not implemented for `i32`, which is required by `W<_>: Downstream`
   |
help: this trait has no implementations, consider adding one
  --> $DIR/false-subtrait-after-inference.rs:15:1
   |
LL | trait Foo {}
   | ^^^^^^^^^
note: required for `W<i32>` to implement `Downstream`
  --> $DIR/false-subtrait-after-inference.rs:13:9
   |
LL | impl<T> Downstream for W<T> where T: Foo {}
   |         ^^^^^^^^^^     ^^^^          --- unsatisfied trait bound introduced here

error: aborting due to 1 previous error; 1 warning emitted

For more information about this error, try `rustc --explain E0277`.
